File: list6/11085 - Back to the 8-Queens.cpp
#include <algorithm>
#include <cstdio>
 
using namespace std;
 
int queens[8];
int path[8];
 
bool isValid(int r, int c) {
    for(int j = 0; j < r; j++) {
        if(c == path[j]) {
            return false;
        }
        
        if(abs(r - j) == abs(c - path[j])) {
            return false;
        }
    }

    return true;
}
 
int search(int n) {
    int total = 100000;
    int i;

    if(n == 8) {
        return 0;
    }
 
    for(i = 0; i < 8; i++)
        if(isValid(n, i)) {
            path[n] = i;

            total = min(total, i == queens[n] ? search(n + 1) : 1 + search(n + 1));
        }
 
    return total;
}
 
int main() {
    int item, j;

    for(item = 1;; item++) {
        for(j = 0; j < 8; j++) {
            if(scanf("%d", &queens[j]) == -1) {
                return 0;
            }

            queens[j]--;
            path[j] = queens[j];
        }
 
        printf("Case %d: %d\n", item, search(0));
    }
}

File: list6/11085 - Back to the 8-Queens.py
# -*- coding: UTF-8 -*-  
queens = []
path = []

def isValid(r, c):
    global queen, path

    for j in range(r):
        if (c == path[j]):
            return False
        
        if (abs(r - j) == abs(c - path[j])):
            return False
    
    return True

def search(n):
    global queen, path
    total = 100000

    if (n == 8):
        return 0
    
    for i in range(8):
        if isValid(n, i):
            path[n] = i

            total = min(total, search(n + 1) if i == queens[n] else 1 + search(n + 1))

    return total


def main():
    global queens, path
    count = 1

    while True:
        try:
            line = input().split()

            queens = []
            path = []

            for item in line:
                queens.append(int(item) - 1)
                path.append(int(item))

            print('Case ' + str(count) + ': ' + str(search(0)))

            count += 1

            
        except(EOFError):
            break
        
if __name__=='__main__':
	main()
